### DISCUSSION TITLE
Pick behavior w.r.t skybox

### DISCUSSION THREAD

**Discussion by achu-topshelf** [google/filament#7140] [Q&A]
Without a skybox, pick works and returns a unique entity, as verified with getId()
When a skybox is set, pick always returns the skybox. Is this unique to my scene (which just loads a glb containing ~10 entities) or do you also see the same behavior (pick always returning skybox).

**Comment by romainguy** [comment]
@pixelflinger

**Reply by pixelflinger** [reply]
Are all your objects transparent by any chance? Transparent objects are not taken into account by picking at the moment.

**Reply by achu-topshelf** [reply]
The objects are not transparent. pick works (in the sense it returns unique entitiyId depending on (x,y) when scene->setSkybox() is commented out. It always returns the skybox when skybox is set, which is the issue.

Another observation (following your comment) is that I had a groundPlane with the material property 'transparentShadow : true'. This resulted in pick returning entityId 0 on touch events in the gaps of the grid (rendered with a shader). I have since commented out this property since I do not need this.

**Reply by romainguy** [reply]
A ground plane with transparent shadows is a special type of entity that would indeed participate in picking. @pixelflinger do we have a way to make a renderable not participate in picking? Layers?

**Reply by pixelflinger** [reply]
No it's very strange, if the only difference is the skybox, I'm not sure what the problem can be.

**Reply by achu-topshelf** [reply]
@pixelflinger I think this is related to my gltf and with hand editing th egltf json I narrowed it down to alphaMode field in the materials section of the gltf (see below)
```
   "materials": [
     {
       "pbrMetallicRoughness": {
         "baseColorFactor": [
           0.913098651791473,
           0.9046611743890203,
           0.9046611743890203,
           1
         ],
         "metallicFactor": 0.5,
         "roughnessFactor": 0.5
-      }
+      },
+      "alphaMode": "BLEND"
```
The above gltf is exported from my scene using a vanilla Threejs GLTFExporter. 
Looking at the code for GLTFExporter I see that alphaMode: blend is set when the material is transparent, so your original hunch is right.
At my end I am trying to figure out where I am setting the material to be transparent when calling the threejs APIs. Will close this if you think there is nothing to be done at the filament side.

**Reply by achu-topshelf** [reply]
Fixed the gltf assets so that the material is not transparent, to the extent possible given the theme. Of course transparent materials with opacity are not picked here. But this is probably ok since there are non-transparent surfaces on the same object that the user can try and click.

The ground plane, with the below material has the following behaviour.
```
material {
    name : Ground,
    blending : transparent,
    doubleSided : true,
    shadingModel : unlit,
    vertexDomain : object,
 ...
```
_A. skybox enabled:_
If blending is commented out, groundplane is picked
If blending is set to 'transparent', groundplane is NOT picked, but skybox is

_B. skybox disabled:_
groundplane is always picked irrespective of whether 'blending' is commented out or not. This is **surprising** if the picking is local to whether the material is transparent or not

Screen looks nicer with skybox enabled and groundPlane blending set to transparent, hence the request.

**Comment by achu-topshelf** [comment]
Thanks @pixelflinger @romainguy and team for the tight library. I have moved on after color matching the ground plane with the skybox.

At this point my interest is theoretical and if you could suggest a pointer to the code where transparency of the material is checked during picking, I could learn something. Feel free to close this discussion too. Thanks again!!
